service/raft/group0_state_machine.cc

```diff
@@ -77,169 +74,53 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
-                co_await write_mutations_to_database(sm._sp, cmd.creator_addr, std::move(chng.mutations));
-                co_await sm._ss.topology_transition(sm._cdc_gen_svc);
-            },
-            [&] (write_mutations& muts) -> future<> {
-                return write_mutations_to_database(sm._sp, cmd.creator_addr, std::move(muts.mutations));
-            }
-            ), cmd.change);
-
-            co_await sm._sp.mutate_locally({std::move(history)}, nullptr);
-        }
+future<> group0_state_machine::apply_command(group0_command cmd, mutation history) {
+    // We assume that `cmd.change` was constructed using group0 state which was observed *after* `cmd.prev_state_id` was obtained.
+    // It is now important that we apply the change *before* we append the group0 state ID to the history table.
+    //
+    // If we crash before appending the state ID, when we reapply the command after restart, the change will be applied because
+    // the state ID was not yet appended so the above check will pass.
+
+    // TODO: reapplication of a command after a crash may require contacting a quorum (we need to learn that the command
+    // is committed from a leader). But we may want to ensure that group 0 state is consistent after restart even without
+    // access to quorum, which means we cannot allow partially applied commands. We need to ensure that either the entire
+    // change is applied and the state ID is updated or none of this happens.
+    // E.g. use a write-ahead-entry which contains all this information and make sure it's replayed during restarts.
+
+    co_await std::visit(make_visitor(
+    [&] (schema_change& chng) -> future<> {
+        return _mm.merge_schema_from(netw::messaging_service::msg_addr(std::move(cmd.creator_addr)), std::move(chng.mutations));
+    },
+    [&] (broadcast_table_query& query) -> future<> {
+        auto result = co_await service::broadcast_tables::execute_broadcast_table_query(_sp, query.query, cmd.new_state_id);
+        _client.set_query_result(cmd.new_state_id, std::move(result));
+    },
+    [&] (topology_change& chng) -> future<> {
+        co_await write_mutations_to_database(_sp, cmd.creator_addr, std::move(chng.mutations));
+        co_await _ss.topology_transition(_cdc_gen_svc);
+    },
+    [&] (write_mutations& muts) -> future<> {
+        return write_mutations_to_database(_sp, cmd.creator_addr, std::move(muts.mutations));
+    }
+    ), cmd.change);
 
-        future<> merge_and_apply() {
-            auto [c, h] = merge();
-            return apply(std::move(c), std::move(h));
-        }
+    co_await _sp.mutate_locally({std::move(history)}, nullptr);
+}
 
-        bool empty() const {
-            return cmd_to_merge.empty();
-        }
+future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merger) {
+    auto [cmd, history] = merger.merge();
+    co_await apply_command(std::move(cmd), std::move(history));
```

Comment:
subjective nit: I would inline `apply_command` here, since it's the only use of it.